python/cudf/cudf/core/buffer/tenable_buffer.py

```diff
@@ -0,0 +1,302 @@
+        )
+
+
+class BufferSlice(TenableBuffer):
+    """A slice (aka. a view) of a tenable buffer.
+
+    Parameters
+    ----------
+    base
+        The tenable buffer this slice refers to.
+    offset
+        The offset relative to the start memory of base (in bytes).
+    size
+        The size of the slice (in bytes)
+    passthrough_attributes
+        Name of attributes that are passed through to the base as-is.
+    """
+
+    def __init__(
+        self,
+        base: TenableBuffer,
+        offset: int,
+        size: int,
+        *,
+        passthrough_attributes: Container[str] = ("exposed",),
+    ) -> None:
+        if size < 0:
+            raise ValueError("size cannot be negative")
+        if offset < 0:
+            raise ValueError("offset cannot be negative")
+        if offset + size > base.size:
+            raise ValueError(
+                "offset+size cannot be greater than the size of base"
+            )
+        self._base = base
+        self._offset = offset
+        self._size = size
+        self._owner = base
+        self._passthrough_attributes = passthrough_attributes
+        base._slices.add(self)
+
+    def __getattr__(self, name):
+        if name in self._passthrough_attributes:
+            return getattr(self._base, name)
+        raise AttributeError(
+            f"{self.__class__.__name__} object has no attribute {name}"
+        )
+
+    def _getitem(self, offset: int, size: int) -> BufferSlice:
+        return BufferSlice(
+            base=self._base, offset=offset + self._offset, size=size
+        )
+
+    @property
+    def exposed(self) -> bool:
+        return self._base.exposed
+
+    def get_ptr(self, mode: str = "write") -> int:
+        if mode == "write" and cudf.get_option("copy_on_write"):
+            self.make_single_owner_inplace()
+        return self._base.get_ptr(mode=mode) + self._offset
+
+    def memoryview(
+        self, *, offset: int = 0, size: Optional[int] = None
+    ) -> memoryview:
+        return self._base.memoryview(offset=self._offset + offset, size=size)
+
+    def copy(self, deep: bool = True) -> BufferSlice:
+        """Return a copy of Buffer.
+
+        What actually happens when `deep == False` depends on the
+        "copy_on_write" option. When copy-on-write is enabled, a shallow copy
+        because a deep copy if the buffer has been exposed. This is because we
+        have no control over knowing if the data is being modified when the
+        buffer has been exposed to third-party.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            The semantics when copy-on-write is disabled:
+                - If True, returns a deep copy of the underlying Buffer data.
+                - If False, returns a shallow copy of the Buffer pointing to
+                  the same underlying data.
+            The semantics when copy-on-write is enabled:
+                - Always a deep copy of the underlying Buffer data.
```

Comment:
I thought with C-o-W, we always get a shallow copy here. (Unless the data are exposed).

Reply:
As far as i understand, from the user's point of view everything is a deep copy ?

Reply:
Ah, I see what you mean. Perhaps something along the lines of:

> Always a deep copy of the underlying Buffer data which is either eager (if the buffer is exposed) or lazy (if not)

Reply:
Maybe that doesn't actually clarify things because we've introduced new terminology.